### ISSUE TITLE
Factors are Silently Converted to Numeric Variables

### ISSUE THREAD

**Issue by DarioS** [dmlc/xgboost#3098]
I'm using the R version of xgboost. It seems to lack support for factors, which every other classifier I've used in R (e.g. classifiers from sparsediscrim and e1071 packages) handles properly.

```
library(xgboost)
datasetObject <- xgb.DMatrix(as.matrix(iris[, 1:4]), label = iris[, "Species"])
trained <- xgb.train(list(eta = 0.5), datasetObject, nrounds = 1)
predicted <- predict(trained, as.matrix(iris[, 1:4]))
str(predicted) # Numeric, even though the class label is a factor.
```

The predictions should be of the same data type as the input labels provided by the user during training.

**Comment by khotilov** [comment]
I'm sorry, things are kept simple here: numeric inputs & numeric outputs. It was like that for awhile, and there's no intention to change. 

Besides, for multiclass classification you would have to set objective to either 'multi:softmax' or 'multi:softprob' and make the class index zero-based, as you may find in the examples in the R help pages.

And you may also use xgboost wrappers in either caret or mlr packages if you look for more "typical" interface.
